doc/sphinx/scripting/python/fontforge.rst

```diff
@@ -511,58 +514,130 @@ Not a very useful example.
-      provide some context for the menu item.
+      When specified this function is called with the same arguments as ``callback``
+      right before the menu or submenu is diplayed. When it returns ``True``
+      the menu item will be enabled and when it returns``False`` it will be
+      disabled. (When ``enable`` is ``None`` the menu item is always enabled.)
 
-   .. object:: which_window
+   .. object:: data
 
-      May be either of the strings ``"Font"`` or ``"Glyph"`` (or the tuple
-      ``("Font", "Glyph")``) and it determines which type of window will have
-      this menu item in its "Tools" menu.
+      ``data`` can be whatever you want; it defaults to ``None``. FontForge
+      passes it to both of the above functions. It can be used to provide
+      context or default arguments for the function (so that one function can
+      be used for multiple menu items.)
 
-   .. object:: shortcut-string
+   .. object:: context
 
-      .. warning:: Deprecated?
+      Currently this can the string ``"Font"``, the string ``"Glyph"``
+      or the tuple ``("Font", "Glyph")``). ``"Font"`` will add the menu item
+      to the FontView "Tools" menu or its submenu, while ``"Glyph"`` will
+      add it to the CharView tools menu or its submenu.
 
-      May be ``None`` if you do not wish to supply a shortcut. Otherwise should
-      be a string like "Menu Name|Cntl-H" (the syntax is defined in the
-      translation section).
+   .. object:: hotkey
 
-   .. object:: submenu-names
+      ``hotkey`` must be either ``None`` or a string in hotkey format,
+      which is the same as the second part of a
+      :ref:`HotKey assignment <HotKeys.hotkeyassign>`.
+      Because hotkeys are a "limited
+      resource" this string is only a `suggestion`; it has no effect
+      when the specified HotKey is already taken. Therefore, before picking
+      a candidate HotKey you should at least verify that it is not already
+      used by the relevant window in FontForge.
 
-      You may specify as many of these as you wish (including leaving them out
-      altogether), this allows you to organize the Tools menu into submenus. (If
-      a submenu of this name does not currently exist, FontForge will create it).
+      Even when the specified HotKey is taken a user can still specify their
+      own in the HotKeys file. You can make this easier to do, now and in the
+      future, by providing the full triplet of names for each "level" using
+      the current interface.
 
-   .. object:: menu-name
+   .. object:: name
 
-      The name that will appear in the menu for this item. This will only affect
-      windows created after this command is executed. Normally the command will
-      be executed at startup and so it will affect all windows.
+      ``name`` can be a string but ideally it is a tuple of three strings
+      ``(localized_name, english_name, identifier_string)`` or of two strings
```

Comment:
Why is it necessary to specify the English name? The hotkey system (questionably) looks for shortcuts based on the untranslated text from the hotkeys file, so I can somewhat understand that, but all of these hotkeys are 'volatile' in that they are registered every time this function is called. Perhaps I'm missing something?

Reply:
Alternatively if it is somehow a requirement for hotkeys, why not just use the identifier (which imo, is what should have been done for all hotkeys; assigning hotkey identifiers instead of relying on untranslated strings)

Reply:
The hotkey specified at registration time is just a suggestion -- it has lowest priority compared with FontForge code hotkeys (which may be added to over time) and user-defined hotkeys.

The English string is for consistency in the user overriding (or just choosing -- I expect most registered menu items will skip the hotkey suggestion) the choice using their `hotkeys` file, which (for reasons that can be questioned but probably aren't changing anytime soon) are always overridden with the English/code-native strings.

Reply:
The reason not to use the identifier is consistency with the existing infrastructure, which doesn't use identifiers anywhere else.

Reply:
Given that the identifier is required to be provided, can that not be used instead?

Reply:
It would be an option to half-overhaul the hotkey system now, but that's not a road I went down. Do you have a suggestion with rough added code points, keeping in mind the Peculiarities of the existing hotkey system (like the initial View prefix)?

Reply:
I can understand elsewhere where there is no explicit need to set an identifier, because the untranslated string is implicitly the identifier, but in this case, you are providing both the untranslated string and an identifier - where the former is only ever used as an identifier anyway.

Reply:
Sorry, seems like I've missed these comments until I refreshed.

I'll take a closer look but I was under the impression that you could just sub the identifier for where it requires the untranslated text.

Reply:
I suppose I'm OK with getting rid of the identifier if necessary. I added it because it seemed preferable to eventually have a menu system based on identifiers and I thought it would be best to encourage plugin developers to add one for when we get there.

Reply:
Yeah - github doesn't seem to be refreshing like it used to.

Reply:
To be clear, I think having identifiers make sense, but I'm proposing to just specify `(text, identifier)`, instead of having `(localised_text, untranslated_text, identifier)`

Reply:
I understand, but I'd prefer (localized_text, untranslated_text) being merged to (text, identifier) being deferred to some future date, given that my adding `identifier` was just future-proofing.

Reply:
It's getting late and I'll spend more time trying this out over the next few days, but my initial impression is that it's pretty buggy.

It seems very hit and miss whether or not a hotkey will 'register', and there's absolutely no feedback if a hotkey is misregistered or anything in the process fails. I've been looking at it with the debugger attached and apparently I need to spend even more time to figure out how to use this, which isn't promising. Apparently a prerequisite to having hotkeys is to have an enable function? But even then half the time it doesn't work.

I've also reinforced my feeling that one of the untranslated/identifier has to go, and I'm very heavily preferring untranslated to go. As to why:

* Localisation is punted to the user - _they_ are the ones providing the localisation, not us. If someone barely knows English and only bothers to have the menu name in their language, why would they bother to translate that to English? More than likely I could see random text being inserted here just to satisfy the requirement
* Even more so, what's the point of providing a translation in two languages if one is never seen to the end user?
* Very little sanitisation is done on the untranslated text. It becomes trivial to write something that would mess up the way hotkey identifiers work (think: adding stray `.`s or `:`s into the string
* At the same time, you _could_ enforce that an identifier is alphanumeric plus maybe underscores/spaces and that seems like a perfectly valid restriction - the same of which cannot be said to what's meant to be a human readable string (even though it's only used for identifier purposes).

Reply:
Am I right that these issues are specific to the hotkey functionality? (So the "buggy" isn't a more general comment about the PR, but the code that is supposed to allow hotkeys to work again (as a suggestion)?)

I didn't have this sort of problem with hotkeys in general in my testing so I'll try to take another look soonish. 

As far at the identifiers-only route goes, I'm not against in principle but is the implementation clear? E.g. how will the menu item wind up with the hotkey printed if there's no association with the menu hierarchy?

Reply:
> Am I right that these issues are specific to the hotkey functionality?

No, just general feeling of registering/use of this functionality overall; I suspect some of it may be long-standing, but it makes testing/using this harder.

Had a bit more time to go through this - bugs (or surprising behaviour):
* (Apparently longstanding): 'Font' context menu items are not added/enabled until a new fontview is created (either by opening a new font or creating a new one)
* The 'Tools' menu is permanently in the 'Enabled' state, even when it has no items
  Old: ![image](https://user-images.githubusercontent.com/5137410/153694960-8b56b4ad-2362-46bd-a9f9-d1f31c30bda6.png)
  New: ![image](https://user-images.githubusercontent.com/5137410/153694978-3f5fbda2-4f13-4082-b95b-5a2c900e5968.png)
* Auto-assignment of mnemonics only happens at the first level, not the submenu
  ![image](https://user-images.githubusercontent.com/5137410/153694549-59810545-e200-468f-8580-3d615204dd75.png)
* If I rerun the same registration code, I end up with two distinct entries instead of overwriting the first. Sample
  ```py
  fontforge.registerMenuItem(print, None, None, "Glyph", "Ctrl+Alt+6", "Sub_Menu", "Print Glyph Name")
  ```
* Assigning a hotkey only works if I omit mnemonic underscores, otherwise no such hotkey is assigned:
  ```py
  fontforge.registerMenuItem(mycb, None, None, "Glyph", "Ctrl+Alt+6", "SubMenu", "Print Glyph Name")
  ```
  (Compare this, which works, to the example above, which does not)
* Non-mnemonic parenthesised character is stripped:
  ```py
  fontforge.registerMenuItem(print, None, None, "Glyph", "Ctrl+Alt+7", "ABC (1)")
  ```
  ![image](https://user-images.githubusercontent.com/5137410/153695892-a29bffdb-46e5-46f7-8b04-51a620a98251.png)


> As far at the identifiers-only route goes, I'm not against in principle but is the implementation clear? E.g. how will the menu item wind up with the hotkey printed if there's no association with the menu hierarchy?

I don't think it's too bad, set the identifier as the untranslated text for the menu item

Reply:
> (Apparently longstanding): 'Font' context menu items are not added/enabled until a new fontview is created (either by opening a new font or creating a new one)

The font menu context problem mirrors most of the rest of the UI: If you change things you typically don't see the result until the next time you open a window.

> The 'Tools' menu is permanently in the 'Enabled' state, even when it has no items

I noticed the tools menu thing. I'll try to fix that. 

> Auto-assignment of mnemonics only happens at the first level, not the submenu

The auto-assignment of mnemonics is intentionally limited to the first level, as noted in the earlier discussions. A sub-menu "collision" is of course possible but for the most part plugin developers will add their own independent submenus (and are encouraged in the plugin documentation to put additions in a submenu). When you know the contents of a submenu there's no possibility of collisions and therefore no need for auto-assignment. 

> If I rerun the same registration code, I end up with two distinct entries instead of overwriting the first. Sample

Having  unregister/reregister work better would be nice but seems unnecessary for this PR given that there's no change.. But I'll look at it.

> Assigning a hotkey only works if I omit mnemonic underscores, otherwise no such hotkey is assigned:

I think is how hotkey assignment works in general. I suppose we could remove underscores in the passed string to improve things, as long as we were confident they aren't being removed in names. (Or I guess we could pass the passed string through one of the `mncopy`s?)

> Non-mnemonic parenthesised character is stripped:

I can try to fix this.

I'm a bit busy before Wednesday and maybe next weekend so I may need until then to look at what I've said I will.

Reply:
> The font menu context problem mirrors most of the rest of the UI: If you change things you typically don't see the result until the next time you open a window.

Sure, I'm just saying that it's not intuitive or immediately apparent. But given it's longstanding it's fine to leave.

> The auto-assignment of mnemonics is intentionally limited to the first level, as noted in the earlier discussions.

My points from my prior comment remain the same

> Having unregister/reregister work better would be nice but seems unnecessary for this PR given that there's no change.. But I'll look at it

I'm not quite sure I understand this, but as is, there is no way to add more than one entry to a submenu, which is broken and needs fixing before this can be merged. It's at least a definite regression from behaviour prior to this PR

> I think is how hotkey assignment works in general. I suppose we could remove underscores in the passed string to improve things, as long as we were confident they aren't being removed in names. (Or I guess we could pass the passed string through one of the mncopys?)

I thought the mnemonic-stripped string was already being passed to the hotkey system, although admittedly I didn't look too deeply into the exact cause

Reply:
> I thought the mnemonic-stripped string was already being passed to the hotkey system, although admittedly I didn't look too deeply into the exact cause

`loadHotkeysFromFile()` leaves `action` as-is and so does `hotkeySetFull()`, as far as I can tell. I think to set hotkeys at present you need to strip the underscores yourself. (I suppose we could also do something more clever than `strcmp()` in `HKActionMatchesFirstPartOf()`.)

Reply:
> The 'Tools' menu is permanently in the 'Enabled' state, even when it has no items

> If I rerun the same registration code, I end up with two distinct entries instead of overwriting the first.

> Non-mnemonic parenthesised character is stripped:

I think I've fixed the first two and greatly (greatly) reduced the chance of the third.

Take a look and also let me know where we are with this after these changes. The window-not-updating and hotkey-file-with-underscores issues seem consistent between normal and python menus. I think I've dealt with what I have to but you might disagree.

> >  As far at the identifiers-only route goes, I'm not against in principle but is the implementation clear? E.g. how will the menu item wind up with the hotkey printed if there's no association with the menu hierarchy?

> I don't think it's too bad, set the identifier as the untranslated text for the menu item

Menus exist in a hierarchy and identifiers don't. (I haven't thought much about submenu-pointing-entries and identifiers. But they're normally not needed or used that way.)

Reply:
> Menus exist in a hierarchy and identifiers don't. (I haven't thought much about submenu-pointing-entries and identifiers. But they're normally not needed or used that way.)

I'm not quite sure I get the implication, but ok.